runtime/integration-tests/src/generic/cases/block_rewards.rs

```diff
@@ -0,0 +1,202 @@
+			CurrencyId::Native,
+			&STAKER.id(),
+			amount,
+		));
+
+		assert_ok!(
+			pallet_rewards::Pallet::<T, BlockRewards>::distribute_reward(200 * CFG, [group_id])
+		);
+
+		assert_eq!(
+			T::Api::list_currencies(RewardDomain::Block, STAKER.id()),
+			vec![CurrencyId::Native]
+		);
+
+		assert_eq!(
+			T::Api::compute_reward(RewardDomain::Block, CurrencyId::Native, STAKER.id()),
+			Some(200 * CFG)
+		)
+	});
+}
+
+crate::test_for_runtimes!([development], collator_list_synchronized);
+fn collator_list_synchronized<T: Runtime>() {
+	RuntimeEnv::<T>::from_parachain_storage(
+		Genesis::default()
+			.add(genesis::balances::<T>(
+				T::ExistentialDeposit::get() + cfg(100),
+			))
+			.add(genesis::invulnerables::<T>(vec![Keyring::Admin]))
+			.add(genesis::session_keys::<T>())
+			.add(genesis::block_rewards::<T>(vec![Keyring::Admin]))
+			.storage(),
+	)
+	.parachain_state_mut(|| {
+		let collators_1 = vec![Keyring::Alice.id(), Keyring::Bob.id()];
+		let collators_2 = vec![
+			Keyring::Charlie.id(),
+			Keyring::Dave.id(),
+			Keyring::Eve.id(),
+			Keyring::Ferdie.id(),
+		];
+
+		// SESSION 0 -> 1;
+		apply_and_check_session::<T>(1, collators_1.clone(), vec![]);
+
+		// SESSION 1 -> 2;
+		apply_and_check_session::<T>(3, collators_2.clone(), vec![]);
+
+		// SESSION 2 -> 3;
+		apply_and_check_session::<T>(7, vec![], vec![Keyring::Alice.id()]);
+
+		assert!(collators_1.iter().all(|c| has_reward::<T>(c)));
+		assert!(collators_2.iter().all(|c| !has_reward::<T>(c)));
+
+		// SESSION 3 -> 4;
+		apply_and_check_session::<T>(6, vec![], vec![]);
+
+		assert!(collators_2.iter().all(|c| has_reward::<T>(c)));
+	});
+}
+
+fn apply_and_check_session<T: Runtime>(
+	collator_count: u32,
+	joining: Vec<AccountId>,
+	leaving: Vec<AccountId>,
+) {
+	for collator in &joining {
+		pallet_collator_selection::Pallet::<T>::register_as_candidate(
+			RawOrigin::Signed(collator.clone()).into(),
+		)
+		.unwrap();
+	}
+
+	for collator in &leaving {
+		pallet_collator_selection::Pallet::<T>::leave_intent(
+			RawOrigin::Signed(collator.clone()).into(),
+		)
+		.unwrap();
+	}
+
+	frame_system::Pallet::<T>::reset_events();
+
+	pallet_session::Pallet::<T>::rotate_session();
+
+	/* TODO: pending to fix. Why try_into() fails getting the reward_event::GroupRewarded?
+
+	// The event exists in this list:
+	dbg!(frame_system::Pallet::<T>::events())
+
+	// But not in in this list (that is the implementation of find_event()),
+	// so try_into returns an Err for it.
+	dbg!(frame_system::Pallet::<T>::events()
+		.into_iter()
+		.rev()
+		.find_map(|record| record.event.try_into().ok())
+		.flatten());
+
+	// But later, if manually I create the event as follows:
+	let e = T::RuntimeEventExt::from(pallet_rewards::Event::<T, BlockRewards>::GroupRewarded {
+		group_id: 1,
+		amount: 2,
+	});
+
+	// And I call try_into(), it works.
+	let re: pallet_rewards::Event<T, BlockRewards> = e.try_into().ok().unwrap();
+	*/
```

Comment:
**99% of developers** can not fix this nor saying what's happening here

cc: @cdamian @wischli @mustermeiszer

Reply:
I don't have a solution at hand right now, would have to debug that myself. Can look into that next week, if desired.

Reply:
Thanks William. My guess is that something weird is happening with the `try_into` when events contain instances. But I've checked all the places where the instance is set, and we always set the correct one for that pallet. 🤔